File: programmers/C++/Level2/숫자 카드 나누기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

int GCD(int a, int b) {
	return b ? GCD(b, a % b) : a;
}

int solution(vector<int> arrayA, vector<int> arrayB) {
	int answer = 0;

	int indexA = 0;
	vector<int> qA = arrayA;
	while (indexA != qA.size() - 1) {
		qA.push_back(GCD(qA[indexA], qA[indexA + 1]));
		indexA += 2;
	}

	int indexB = 0;
	vector<int> qB = arrayB;
	while (indexB != qB.size() - 1) {
		qB.push_back(GCD(qB[indexB], qB[indexB + 1]));
		indexB += 2;
	}

	int A = qA[indexA]; // arrayA의 최대공약수
	int B = qB[indexB]; // arrayB의 최대공약수

	if (all_of(arrayB.begin(), arrayB.end(), [&](int x) { return x % A; }))
		answer = max(A, answer);
	if (all_of(arrayA.begin(), arrayA.end(), [&](int x) { return x % B; }))
		answer = max(B, answer);

	return answer;
}

int main() {
	cout << solution({ 14, 35, 119 }, { 18, 30, 102 });
	return 0;
}

File: programmers/Java/Level2/숫자 카드 나누기.java
import java.util.ArrayList;
import java.util.Arrays;

class Solution {
    public int solution(int[] arrayA, int[] arrayB) {
        int answer = 0;

        ArrayList<Integer> qA = new ArrayList<>();
        ArrayList<Integer> qB = new ArrayList<>();
        for (int n : arrayA)
            qA.add(n);
        for (int n : arrayB)
            qB.add(n);

        int indexA = 0;
        while (indexA != qA.size() - 1) {
            qA.add(GCD(qA.get(indexA), qA.get(indexA + 1)));
            indexA += 2;
        }

        int indexB = 0;
        while (indexB != qB.size() - 1) {
            qB.add(GCD(qB.get(indexB), qB.get(indexB + 1)));
            indexB += 2;
        }

        int A = qA.get(indexA); // arrayA의 최대공약수
        int B = qB.get(indexB); // arrayB의 최대공약수

        if (Arrays.stream(arrayB).allMatch(x -> x % A != 0))
            answer = Math.max(A, answer);
        if (Arrays.stream(arrayA).allMatch(x -> x % B != 0))
            answer = Math.max(B, answer);

        return answer;
    }

    public int GCD(int a, int b) {
        return b != 0 ? GCD(b, a % b) : a;
    }
}